### ISSUE TITLE
Add ability to import user credentials

### ISSUE THREAD

**Issue by aeneasr** [ory/kratos#605]
**Is your feature request related to a problem? Please describe.**

It should be possible to import user/identity credentials when creating identities using the `POST /identities` Admin API.

**Describe the solution you'd like**

There needs to be some type of value (or a dedicated API?) which allows the client to specify the credential. For the credential type, we should be able to work with passwords hashed with

- [ ] BCrypt;
- [ ] Argon2*;
- [ ] PBKDF1/2.

We could also allow other algorithms, such as SHA with rounds or similar methods. We should probably have a list of software platforms and their hashing algorithms to decide what to support. Custom or insecure mechanisms (salted md5?) should probably not be supported in order to keep the code base tidy.

It should also be possible to import OIDC linked accounts.

The idea would be to specify a payload, for example:

```
{
  "traits": ...
  "import_credentials": [
    {
      "type": "password",
      "hash": "...",
      "algorithm": "bcrypt"
    },
    {
      "type": "oidc",
      "provider": "google",
      "user_id": "123456"
    }
  ]
}
```

On the next login, the password would then be converted to a secure Argon2 hash and the original hash will be removed from the database.

**Describe alternatives you've considered**

Instead of providing the key `import_credentials` there would be a collection of endpoints at `/identities/<identity>/credentials` which would allow for PUT, POST, DELETE.

**Comment by aeneasr** [comment]
@Jasrags do you have any preference regarding the API? We're currently thinking about having a dedicated endpoint for that which would mean that you need to do two requests. One to `POST /identities` and then one to `POST /identities/<identity>/credentials`. We'd probably add a CLI helper to help with this.

**Comment by Jasrags** [comment]
@aeneasr I think that could work. I don't have any preferences at the moment.

**Comment by aeneasr** [comment]
Sounds good @Jasrags if you have any other troubles with migrating over to ORY Kratos let me know :)

**Comment by zcmack** [comment]
i'll pile on and say that i have a very, very similar use case and that having a dedicated endpoint / two requests would be just fine. this may be out of scope for this issue, but I figure I'll share for context: in an evaluation of FusionAuth this behaved similarly and worked very well, I was able to use their SDK to easily script a migration however the system would not allow for users to have duplicate email addresses by design. this doesn't work so well for a large public site in which having multiple accounts bound to a single email address is perfectly allowable.

**Comment by ykuzma1** [comment]
Also wanted to chime in that this would be helpful in our use case. We are looking to pre-create a user with the POST /identities endpoint so that we can set various user properties prior to their first login using a custom admin interface.

Current behavior: First POST a new identity. Have that same OIDC user login. The user gets redirected to the registration flow with the following error `An account with the same identifier (email, phone, username, ...) exists already.`. Trying to `Continue with Google` brings you back to the Google OAuth screen which then brings you back to the registration flow and error in an endless loop.

**Comment by zepatrik** [comment]
Can't you send your users to the account recovery flow instead of registration? You can even send them the recovery link by getting it through the admin API (see https://www.ory.sh/kratos/docs/reference/api#create-a-recovery-link).

**Comment by lozybean** [comment]
Should consider compare encoded hash in [PHC string format](https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md).

In `golang.org/x/crypto/bcrypt`, there is no way to generate encoded hash with salt, nor compare hash and password in PHC string format, and `bcrypt.bcrypt` is private too.

**Comment by jandd** [comment]
Open Source implementations for password upgrades exist in i.e.

- Django https://docs.djangoproject.com/en/3.1/topics/auth/passwords/#wrapping-password-hashers
- Spring Security https://docs.spring.io/spring-security/site/docs/5.4.2/api/org/springframework/security/crypto/password/DelegatingPasswordEncoder.html

Maybe these can give a good starting point for the password upgrade part of this feature.

I think for those legacy systems that have even worse password hashing it would be cool to have a good guide how to implement an own hash implementation. The password hash prefix that is available in the Spring security world where all but the default password hash are prefixed with `{hashtype}` might be a good idea too.

**Comment by jandd** [comment]
> Can't you send your users to the account recovery flow instead of registration? You can even send them the recovery link by getting it through the admin API (see https://www.ory.sh/kratos/docs/reference/api#create-a-recovery-link).

that is a very good idea indeed. Much better then carrying the old password hashes arround.

**Comment by realStandal** [comment]
I have a use-case which I have yet to implement but have none-the-less been working through and think I've ran into something worth noting here for y'all. It's on-subject with the idea of credentials being set during the creation of an identity via the `/identities/` endpoint; I did not see any other issue describing this as a feature but this seems to be that place (I'm seeing now a dupe was amidst, c'est la vie).

I'm exploring using Kratos to manage identities in a multi-tenant environment. I plan to deploy it [as suggested](https://www.ory.sh/kratos/docs/guides/multi-tenancy-multitenant/), and have ran into no issues with this plan so far. In addition, I'm planning to take full advantage to Kratos' claim of treating identities incredibly arbitrarily (it can be a human, IoT device, etc.); again, no issues there. In the end, tenants will have full control over the management (and life-cycle, so to speak) of their Identities. Registration for a particular tenant's instance will be disabled.

In terms of managing an identity in this context, it would be beneficial for the credential of an identity to be set during its creation using the `/identities/` end-point in a single request.

Account's used by human's can take advantage of the recovery flow once created, as suggested:

> Can't you send your users to the account recovery flow instead of registration? You can even send them the recovery link by getting it through the admin API (see https://www.ory.sh/kratos/docs/reference/api#create-a-recovery-link).

But for IoT devices, how are they to "reset" their password? For this, lets consider it'd be cumbersome and unnecessarily time-consuming to associate each device with an E-Mail (likely non-uniquely) to force it to use reset behavior.

I'm sure I can (and will need to) make sacrifices and do a bit of leg work to get to my outcome, but I just wanted to highlight this use-case in-case it was beneficial. All-in-all, having the ability to set an identities credentials, using the admin endpoints, would be very beneficial. Letting them be set during creation would be ideal; but even having the second end-point would simply require the additional UI-step of creating the Identity then setting its credential's.

First interaction with y'all as the maintainer(s), lovely software thank you :-)

**Comment by aeneasr** [comment]
@realStandal thank you for your thoughts. We definitely want to add the ability to specify and import passwords. I think that for IoT it would make sense to have some other type of authentication (e.g. a private key) as the notion of a password does not really fit there!

What type of IoT are you looking at?

**Comment by realStandal** [comment]
@aeneasr ambiguous to the kind at the moment. Id planned for the key-based once it was supported, nice to know the ability to do so is there, though :) Is y'all's documentation outdated and it _is_ supported? No difference to me atm either way.

**Comment by aeneasr** [comment]
Importing is not yet implemented, scheduled for 0.7 which lands after MFA and flow refactoring. If you'd like to contribute however, let's chat :)

**Comment by BlueHotDog** [comment]
Hey, is there any workaround this? we're now exploring moving from our home-brew solution to something more robust but we already have users

**Comment by pommelinho** [comment]
@aeneasr  are you also planning to be able to migrate HMAC? Will you support SHA-1?

**Comment by aeneasr** [comment]
You mean importing SHA1 hashed passwords?

**Comment by pommelinho** [comment]
Yes, and additionally are you willing to support HMAC-SHA-1 or HMAC-SHA-256?
And will you also support when combinations are used, so e.g. `bcrypt(HMAC(SHA-256(password), pepper))`?

**Comment by aeneasr** [comment]
We would like not to support all the variants of sha/scrypt/bcrypt/sha2/sha3/… and their combinations as the complexity will eventually explode. Instead we’ll try to cover the major hashing algorithms and frameworks (e.g. PHP style hashing). SHA1 for hashing for example is something I haven’t seen in almost a decade ;)

**Comment by pommelinho** [comment]
Ok thanks for your reply.

However I am still wondering if you will support peppers to prevent dictionary attack when the db gets leaked? See [here](https://security.stackexchange.com/questions/21263/how-to-apply-a-pepper-correctly-to-bcrypt)

In my eyes this would add a significant security improvement. Or do you think plain bcrypt is sufficient?

I

**Comment by aeneasr** [comment]
BCrypt does not need any salting or pepper as it is built into the algorithm. Do you have concrete use cases for BCrypt(Hmac(SHA())) or similar variants (e.g. Laravel uses this pattern and you want to import users) or are these just ideas? If it is the second I would suggest to focus on concrete use case concerns for importing password hashes from other systems and keep potential hashing variations away from this project for now :)

**Comment by pommelinho** [comment]
The use case is the different migration possibilities  from Hmac(SHA()) to bcrypt (see [here ](https://youtu.be/OQD3qDYMyYQ?t=619) we are facing. While doing this we also want to be safe we are compatible to Kratos when import is possible.
The solutions we have are shown [here](https://youtu.be/OQD3qDYMyYQ?t=2260) . Solution 1 here suggests to use a pepper along with bcrypt. However this is only necessary to be safe from "hash shucking" when you are using rehashing.  Rehashing is preferable when you want to update hash algo without forcing the user to reenter password.

So for me it's ok when you are saying you are using bcrypt without any pepper. Then we need to creates new hashs for user passwords with bcrypt to be "Kratos conform", which means the need to reenter passwords

**Comment by aeneasr** [comment]
Wow, interesting talk! Thank you :)

 I’ll give it a full listen and come back to this issue!

**Comment by g13013** [comment]
this is much needed !

**Comment by harnash** [comment]
I have this issue on my radar too. What I can try to do relatively quickly is to add option to import credentials in a format that is currently known to Kratos (currently supported hashers) without any hash wrapping and auto upgrading. This will be the first step and automatic upgrade will follow. I can start working on that probably in a week or so.

**Comment by aeneasr** [comment]
That would be awesome! How would we handle verified addresses? Not sure if we support setting those right now

**Comment by harnash** [comment]
Similar as it is currently returned in response? So we have something like this:
```json
{
"schema_id": "string",
"traits": {},
"verifiable_addresses": [
    {
        "status": "string",
        "value": "string",
        "verified": true,
        "verified_at": "2019-08-24T14:15:22Z",
        "via": "string"
    }
]
"credentials": {
        "type": "string",
        "config": {}
     }
}
```
I'm not 100% sure yet since we would need to support passwords and OIDC so that might change when I get my hands on it. Probably we would need to support `credentials` as an array to allow IODC and password at the same time.

**Comment by aeneasr** [comment]
Yeah that makes sense. I believe credentials are an array!

**Comment by g13013** [comment]
@aeneasr why just not allowing for a configurable external url that would be used only once when the user logs in the first time it logs in:

consider this:

```
flows:
    login:
        externalPasswordVerifier: # used only when the password is not found for the identity
             url: https://....
             method: post
            // auth config
 ```
 
This way Kratos would greatly facilitate account importing and support all legacy password encryption:

so to import an account a service would simply create accounts beforehand and then provide and url that checks passwords against legacy database.

**Comment by lanphan** [comment]
Hi @aeneasr ,

I'd like to introduce my new use-case for this feature, but this is a very simple use-case: I just want to add a list of (user+pass) into kratos in which password will be encoded by Argon2 directly from Kratos and doesn't have any migration step from the old password to the new one, is there any work-around for my use-case?

**Comment by aeneasr** [comment]
That's currently not possible unfortunately

**Comment by BlueHotDog** [comment]
Hi, so currently the only way to start using Kratos is if we start a project without any users?
I've read the thread but i just trying to make sure i understand that there's no way to migrate to Kratos for an existing system(We're using Bcrypt and should be fairly aligned with what Kratos needs)

**Comment by aeneasr** [comment]
You can import users at the moment if it is ok for you to send them links which require them to set up a new password.

It is not possible to import user credentials at the moment!

**Comment by seremenko-wish** [comment]
Can we work on functionality to import user password hashes compatible with hashing algorithms that are supported by Kratos now? So if I have a user base with passwords hashed by argon2 or bcrypt, I could import them into Kratos.

**Comment by snikch** [comment]
We're starting to look at what's required to move our user base out of a large Magento installation. Magento uses a "wrapping" style system for their password hashes that allows them to upgrade the algorithm that is used by wrapping the previous hash in the new algorithm, described in [this doc](https://devdocs.magento.com/guides/v2.4/config-guide/secy/hashing.html).

When it comes to implementing credential importing, I'd be happy to do the work on implementing a Magento compatible pipeline.

I personally think this is going to be a massive win for Kratos as I would imagine without a full migration strategy it's an immediate show stopper for any organisation looking to replatform without having to reset user passwords.

**Comment by aeneasr** [comment]
That is great! So let's start with a design document to address this. I think we need these things in place:

- Ability to import generic password hashes (e.g. BCrypt, PKBDF2, ...) using a standardized hash format (like the one we have for PKBDF2)
- An REST API outline (with payloads and paths) for implementing generic password hashes
- An REST API outline for importing dedicated imports for e.g. magento

Who would be open to start such a doc? I think that once we agree on the spec implementation is very straight forward and fast!

**Comment by snikch** [comment]
I'm happy to start on this, I can find some time in the next two weeks. Do you have some examples of previous docs like this so I know approx how you'd like it formatted?

**Comment by aeneasr** [comment]
Unfortunately not, it’s a new process! The template has a layout and explanations which hopefully are helpful :)

**Comment by snikch** [comment]
Sweet, I've chucked a card into our backlog and will find some time in the next two weeks.

**Comment by mcjimenez** [comment]
We have the same problem where we want to start using hydra and kratos to authenticate users, but we already have a user database and we want to make the move as seamless and transparent for users as possible. And that includes being able to use their existing passwords.

For us most of the features like password upgrade paths and so on discussed on this thread are nice to have, but we don't need it right now. Our problem is also easier since our existing user database already uses bcrypt.

With that in mind, we've implemented a change to the POST /identities API to be able to include a hashed (or cleartext) password, and a set of Verifiable Addresses (because our users already have their addresses verified). The (draft) PR with our current set of changes is on https://github.com/ory/kratos/pull/1963

That's currently a draft PR because I don't know if you want to try to land that as a first approach and then add the rest of the functionality. If you think it's worth finishing this to land it, I can move it to a full PR and start the reviewing process.

**Comment by harnash** [comment]
@mcjimenez we have something similar in tests right now https://github.com/ory/kratos/pull/1963 and perhaps we could have come up with a solution that would work for both use cases (ping @slayful).

**Comment by snikch** [comment]
So upon further investigation, it turns out the majority of users in our dataset in Magento have Argon2 hashed passwords that we can import into Kratos. This means that the only blocker for us is the actual password import step which looks to be close to being solved via #1963. I'm happy to help there if anything is blocking @mcjimenez?

**Comment by aeneasr** [comment]
Nice! :) The biggest blocker for merging PR is the CLA. Not sure if @mcjimenez wants to sign it but if lot it would mean that we need a new PR with original changes :/

**Comment by g13013** [comment]
@aeneasr what do you think of my proposal ?

https://github.com/ory/kratos/issues/605#issuecomment-894845348

we are stuck by this issue and https://github.com/ory/kratos/issues/970

as it takes a lot of time we unfortunately think about abandoning kratos as a solution for us mainly because it seems to be made for now projects and does not take into account existing projects.

**Comment by kszafran** [comment]
@g13013 Have you considered migrating existing users by inserting directly into Kratos' database? I haven't worked out the details yet, but that's how I'm planning to move our current users to Kratos.

**Comment by g13013** [comment]
@kszafran yes, it doesn't work for us for 2 reasons:

- Does not accept to create identities with their password, so the user has to pass by the `self-service` to provide the password
- Password policies are too strict.

**Comment by kszafran** [comment]
I can't comment on password policies, but for the first bullet point, can't you insert user credentials directly into the DB? I imagine calling `POST /identities` is going to be done by some backend job/process, which could imaginably also have access to Kratos' database.

**Comment by aeneasr** [comment]
There’s a PR open for credentials but it looks abandoned. From our team noone has time to work on this right now. But the good thing about open source is that everyone can step up and contribute :) So why not make a PR if this blocks you?

**Comment by mcjimenez** [comment]
I'm sorry for having gone dark on this. We're still working on this PR but we ran into some problem internally with signing the CLA, and we're trying to get that untangled

**Comment by zcmack** [comment]
@mcjimenez any updates on this? i'd be glad to create a new PR that is "inspired" by your work for the sake of resolving this issue. i wouldn't want to do this without your permission though, especially if its just corporate red tape that must be resolved.

**Comment by mcjimenez** [comment]
I'm sorry for the delay, I was on a Christmas break. It's mostly corporate red tape, which I hope will be fixed soon (as in no longer than this week). I'll get back to you if that's not the case
